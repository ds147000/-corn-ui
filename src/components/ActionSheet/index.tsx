/* eslint-disable import/export */

import ActionSheet  from './actionSheet'

export { ActionSheetItem } from './item'
export default ActionSheet
