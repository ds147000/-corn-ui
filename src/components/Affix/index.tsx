/* eslint-disable no-unreachable */
/* eslint-disable no-magic-numbers */
// #if _APP === 'weapp'
import Taro, { usePageScroll } from '@tarojs/taro'
// #endif
import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react'
import { View } from '@tarojs/components'
import ClassNames from 'classnames'
import { ViewProps } from '../../types/View'
import { judge } from './utils/judge'

export interface AffixProps extends ViewProps {
  position?: 'top' | 'bottom' | 'right' | 'left'
  onChange?(fixed: boolean): void
}

const Affix: React.FC<AffixProps> = ({ children, position = 'top', onChange, className, ...props }) => {
  const [ fixed, setFixed ] = useState(false)
  const [ boxStyle, setBoxStyle ] = useState<React.CSSProperties>({})
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const ref = useRef<any>()

  const oldFixed = useRef<boolean>(false)

  const _class = useMemo(() => {
    return ClassNames('corn-affix', fixed && `corn-affix-${position}`, className)
  }, [ fixed, position, className ])

  const changeFixed = useCallback((value: boolean): void => {
    setFixed((state) => {
      oldFixed.current = state
      return value
    })
  }, [])

  // #if _APP === 'weapp'
  const system = useRef(Taro.getSystemInfoSync())
  const timer = useRef<NodeJS.Timeout>()

  const checkFlexd = (): void => {
    const node = ref.current

    const rootRect = {
      top: 0,
      left: 0,
      bottom: system.current.windowHeight,
      right: system.current.windowWidth
    } as DOMRectReadOnly

    Taro.createSelectorQuery()
      .select('#' + node.uid)
      .fields({ size: true, rect: true }, (res) => {
        if (!res) return

        setBoxStyle({ height: res?.height, width: res?.width })
        const result = judge(res as unknown as DOMRect, rootRect, position, 0)
        changeFixed(result)
      })
      .exec()
  }

  usePageScroll(() => {
    if (timer.current) clearTimeout(timer.current)
    timer.current = setTimeout(() => checkFlexd(), 10)
  })

  useEffect(() => {
    return (): void => {
      if (timer.current) clearTimeout(timer.current)
    }
  }, [])

  // #else
  useEffect(() => {
    // eslint-disable-next-line no-unreachable
    const el: HTMLDivElement = ref.current
    const rootBounds = {
      x: 0,
      y: 0,
      width: window.innerWidth,
      height: window.innerHeight,
      top: 0,
      left: 0,
      bottom: window.innerHeight,
      right: window.innerWidth
    } as DOMRectReadOnly

    setBoxStyle({ height: el.offsetHeight, width: el.offsetWidth })

    function handleScroll(): void {
      const item = el.getBoundingClientRect()
      const result = judge(item, rootBounds, position, 0)
      changeFixed(result)
    }
    handleScroll()
    window.addEventListener('scroll', handleScroll)
    return (): void => {
      window.removeEventListener('scroll', handleScroll)
    }
  }, [ position, changeFixed ])
  // #endif

  useEffect(() => {
    if (oldFixed.current !== fixed) {
      oldFixed.current = fixed
      onChange?.(fixed)
    }
  }, [ fixed, onChange ])

  return (
    <View style={boxStyle} ref={ref}>
      <View className={_class} {...props} >
        {children}
      </View>
    </View>
  )
}

export default Affix
