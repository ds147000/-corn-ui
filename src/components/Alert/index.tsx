import React from 'react'
import { View } from '@tarojs/components'
import classNames from 'classnames'
import { ViewProps } from '../../types/View'
import Icon from '../Icon'

const Alter: React.FC<ViewProps> = ({ children, className, ...props }) => {
  const _class = classNames(className, 'corn-alert', 'corn-f')
  return (
    <View className={_class} {...props} >
      <Icon name="error" className="icon" />
      {children}
    </View>
  )
}

export default Alter
