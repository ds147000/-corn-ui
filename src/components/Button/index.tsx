import React, { useMemo } from 'react'
import ClassNames from 'classnames'
import { Button as TaroButton } from '@tarojs/components'
import Link, { LinkProps } from '../Link'
import { ITouchEvent } from '../../types'

/** 按钮大小，max: 最大， big: 超大，large：大，middle：中，small：小，mini */
export type ButtonSize = 'max' | 'big' | 'large' | 'middle' | 'small' | 'mini'
export type ButtonType = 'primary' | 'warn' | 'error' | 'link' | 'pop' | 'default' | 'light' | 'stop'

export interface ButtonProps extends LinkProps {
  /** 按钮大小，max: 最大， big: 超大，large：大，middle：中，small：小，mini */
  size?: ButtonSize
  /** 按钮风格类型 */
  type?: ButtonType
  /** 图标 */
  icon?: React.ReactNode
  /** 将按钮宽度调整为其父宽度的选项	 */
  block?: boolean
  /** 镂空风格 */
  ghost?: boolean
  /** 是否禁用 */
  disabled?: boolean
  /** 跳转链接，填写启用A链接行为 */
  to?: string
  /** 表单点击行为 */
  formType?: 'submit' | 'reset'
  /** 开启固定宽度 */
  auto?: boolean
  /** 点击事件 */
  onClick?(e: ITouchEvent): void
}

const Button: React.FC<ButtonProps> = ({
  size = 'middle',
  type = 'primary',
  block,
  ghost,
  disabled,
  to,
  onClick,
  icon,
  children,
  className,
  formType,
  auto,
  ...props
}) => {
  const isIcon = icon !== undefined
  const isChildren = children !== undefined

  const _class = useMemo(() => {
    return ClassNames(
      'corn-btn',
      'corn-if corn-ac corn-jc',
      `corn-btn-${size}`,
      `corn-btn-${type}`,
      {
        'corn-btn-block': block,
        'corn-btn-ghost': ghost,
        'corn-btn-icon': isIcon && !isChildren,
        'corn-btn-child-icon': isIcon && isChildren,
        'corn-btn-disabled': disabled,
        'corn-btn-auto': auto
      },
      className
    )
  }, [ size, type, block, ghost, disabled, className, auto, isIcon, isChildren ])

  const _onClick = (e: ITouchEvent): void => {
    if (disabled) return

    onClick?.(e)
  }



  if (to) {
    return (
      <Link {...props} target="Text" to={to} className={_class} onClick={_onClick} disabled={disabled} type="default" >
        {children}
        {icon}
      </Link>
    )
  }

  return (
    <TaroButton {...props} formType={formType} className={_class} onClick={_onClick} >
      {children}
      {icon}
    </TaroButton>
  )
}

export default Button
