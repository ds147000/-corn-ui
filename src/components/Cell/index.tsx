import React, { useCallback } from 'react'
import ClassNames from 'classnames'
import { View, Text } from '@tarojs/components'
import { ViewProps } from '@tarojs/components/types/View'
import Icon from '../Icon'
import Link, { LinkProps } from '../Link'
import CellList from './list'

export interface CellProps extends ViewProps, LinkProps {
  /** 标题 */
  label?: string | React.ReactNode
  /** 内容 */
  value?: string | React.ReactNode
  /** 内容为空占位提示 */
  placeholder?: string | React.ReactNode
  /** 是否显示箭头 */
  arrow?: boolean
  /** 尾缀内容 */
  suffix?: React.ReactNode | string
  /** 禁用 */
  disabled?: boolean
  /** 跳转链接，填写启用A链接行为。同时会显示箭头 */
  to?: string
}

type CellView =  { List: typeof CellList }



const Cell: React.FC<CellProps> & CellView = ({
  label, value, arrow, suffix,
  disabled, to, children, onClick,
  className, placeholder, ...props
}) => {

  const _onClick = useCallback((event): void => {
    if (disabled) return
    onClick?.(event)
  }, [ onClick, disabled ])

  const _props: ViewProps = {
    className: ClassNames('xrk-cell xrk-f xrk-ac', disabled && 'xrk-cell-disable', className),
    onClick: _onClick,
    ...props
  }

  const showArrow = (arrow || Boolean(to)) && arrow !== false
  const showPlaceholder = Boolean(value) === false && Boolean(children) === false

  const content = (
    <>
      <View className="xrk-cell-label">{label}</View>
      <View className="xrk-cell-value">
        {value}{children}
        {showPlaceholder && <Text className="xrk-cell-placeholder">{placeholder}</Text>}
      </View>
      <View className="xrk-cell-suffix xrk-f xrk-ac">
        {suffix}
        {showArrow && <Icon name="big-right" data-testid="arrow" />}
      </View>
    </>
  )

  if (to)
    return (<Link {..._props} to={to} disabled={disabled} target="View" >{content}</Link>)

  return (<View {..._props} >{content}</View>)
}

Cell.List = CellList

export default Cell
