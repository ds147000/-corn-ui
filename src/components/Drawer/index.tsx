import React, { useCallback, useEffect, useMemo, useState } from 'react'
import classNames from 'classnames'
import { View } from '@tarojs/components'

type AnimateStatus = 'show' | 'hide' | 'over'

export interface DrawerProps {
  /** 是否显示 */
  visible: boolean
  /** 弹出的方向，可以设置左右上下中间, 默认 bottom */
  position?: 'left' | 'right' | 'bottom' | 'top' | 'center'
  /** 是否显示遮罩层 */
  mask?: boolean
  /** 点击蒙层是否允许关闭 */
  maskClosable?: boolean
  /** 关闭回调 */
  onClose?(): void
  /** 完全隐藏回调 */
  onHide?(): void
}

const Drawer: React.FC<DrawerProps> = ({
  visible,
  position = 'bottom',
  mask = true,
  maskClosable = true,
  onClose,
  onHide,
  children
}) => {
  const [ status, setStatus ] = useState<AnimateStatus>('over')
  const _class = useMemo(() => {
    return classNames(
      'xrk-drawer-body',
      status !== 'over' && `xrk-drawer-${position}-${status}`,
      `xrk-drawer-${position}`
    )
  }, [ position, status ])

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const onAnimationEnd = useCallback((event: any) => {
    const animationName: string = event.animationName ||
      event.detail?.animationName ||
      event.nativeEvent?.animationName ||
      event.target.className

    // eslint-disable-next-line no-magic-numbers
    if (animationName.toLocaleUpperCase().indexOf('HIDE') !== -1) setStatus('over')
  }, [])

  const onClickMask = (): void => {
    if (maskClosable) onClose?.()
  }

  useEffect(() => {
    if (visible === true && status !== 'show')
      setStatus('show')

    else if (visible === false && status === 'show')
      setStatus('hide')

  }, [ visible, status ])

  if (visible === false && status === 'over') {
    onHide?.()
    return null
  }

  return (
    <>
      {mask && (
        <View
          className={`xrk-drawer-mask xrk-drawer-mask-${status}`}
          data-testid="mask"
          onClick={onClickMask}
        />
      )}
      <View className={_class} onAnimationEnd={onAnimationEnd} data-testid="body" >
        {children}
      </View>
    </>
  )
}

export default Drawer
