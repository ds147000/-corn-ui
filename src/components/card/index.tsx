/*
 * @Description:
 * @Author: zhoulong.yang
 * @Date: 2021-06-07 12:03:34
 * @LastEditors: zhoulong.yang
 * @LastEditTime: 2021-06-07 15:43:55
 */
import React from 'react'
import { View, Text } from '@tarojs/components'

const Card: React.FC = () => {
  return (
    <View>
      <Text>卡片</Text>
    </View>
  )
}

export default Card
