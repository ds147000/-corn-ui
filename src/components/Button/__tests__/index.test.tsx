/* eslint-disable no-magic-numbers */
import React from 'react'
import { fireEvent, render, waitFor } from '@testing-library/react'
import { View, Text } from '@tarojs/components'
import { BrowserRouter, Switch, Route } from 'react-router-dom'
import Button from '../index'

describe('<Button />', () => {
  test('基本使用', () => {
    const screen = render(<Button />)
    expect(screen.container).toMatchSnapshot()
  })

  test('标题：向日葵妈妈UI', () => {
    const screen = render(<Button>向日葵妈妈UI</Button>)
    screen.getByText('向日葵妈妈UI')
    expect(screen.container).toMatchSnapshot()
  })

  test('复杂标题：节点+文字', () => {
    const screen = render(
      <Button>
        <View>
          <Text>我是图标</Text>
        </View>
        <Text>向日葵妈妈UI</Text>
      </Button>
    )
    screen.getByText('向日葵妈妈UI')
    screen.getByText('我是图标')
    expect(screen.container).toMatchSnapshot()
  })

  test('点击事件', () => {
    const onClick = jest.fn()
    const screen = render(<Button onClick={onClick}>向日葵妈妈UI</Button>)
    fireEvent.click(screen.getByText('向日葵妈妈UI'))
    expect(onClick).toHaveBeenCalledTimes(1)
  })

  test('禁用', () => {
    const onClick = jest.fn()
    const screen = render(<Button onClick={onClick} disabled >向日葵妈妈UI</Button>)
    fireEvent.click(screen.getByText('向日葵妈妈UI'))
    expect(onClick).toHaveBeenCalledTimes(0)
  })

  test('size', () => {
    const screen = render(
      <View>
        <Button size="max">向日葵妈妈UI</Button>
        <Button size="big">向日葵妈妈UI</Button>
        <Button size="large">向日葵妈妈UI</Button>
        <Button size="middle">向日葵妈妈UI</Button>
        <Button size="small">向日葵妈妈UI</Button>
        <Button size="mini">向日葵妈妈UI</Button>
      </View>
    )
    expect(screen.container).toMatchSnapshot()
  })

  test('type', () => {
    const screen = render(
      <View>
        <Button type="error">向日葵妈妈UI</Button>
        <Button type="link">向日葵妈妈UI</Button>
        <Button type="pop">向日葵妈妈UI</Button>
        <Button type="primary">向日葵妈妈UI</Button>
        <Button type="warn">向日葵妈妈UI</Button>
      </View>
    )
    expect(screen.container).toMatchSnapshot()
  })

  test('icon', () => {
    const screen = render(<Button icon={<View>我是图标</View>}>向日葵妈妈UI</Button>)
    screen.getByText('我是图标')
    expect(screen.container).toMatchSnapshot()
  })

  test('block', () => {
    const BlockScreen = render(<Button block >向日葵妈妈UI</Button>).container.innerHTML.toString()
    const DefualtScreen = render(<Button block={false} >向日葵妈妈UI</Button>).container.innerHTML.toString()

    expect(BlockScreen).not.toBe(DefualtScreen)
  })

  test('ghost', () => {
    const BlockScreen = render(<Button ghost >向日葵妈妈UI</Button>).container.innerHTML.toString()
    const DefualtScreen = render(<Button ghost={false} >向日葵妈妈UI</Button>).container.innerHTML.toString()

    expect(BlockScreen).not.toBe(DefualtScreen)
  })

  test('href', async () => {
    const screen = render(
      <BrowserRouter>
        <Switch>
          <Route>
            <Button to="/home" >向日葵妈妈UI</Button>
          </Route>
        </Switch>
      </BrowserRouter>
      )
    await waitFor(() => expect(screen.container.querySelector('a').href).toBe('http://localhost/home'))
    expect(screen).toMatchSnapshot()
  })

  test('固定宽带', () => {
    const screen = render(<Button auto />)
    expect(screen.container).toMatchSnapshot()
  })

})
