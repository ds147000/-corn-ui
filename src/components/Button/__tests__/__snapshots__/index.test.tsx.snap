// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> href 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <a
        class="corn-link corn-link-default corn-btn corn-if corn-ac corn-jc corn-btn-middle corn-btn-primary"
        href="/home"
      >
        @CornUI
      </a>
    </div>
  </body>,
  "container": <div>
    <a
      class="corn-link corn-link-default corn-btn corn-if corn-ac corn-jc corn-btn-middle corn-btn-primary"
      href="/home"
    >
      @CornUI
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<Button /> icon 1`] = `
<div>
  <button
    class="corn-btn corn-if corn-ac corn-jc corn-btn-middle corn-btn-primary corn-btn-child-icon"
    type="button"
  >
    @CornUI
    <div
      class=""
    >
      我是图标
    </div>
  </button>
</div>
`;

exports[`<Button /> size 1`] = `
<div>
  <div
    class=""
  >
    <button
      class="corn-btn corn-if corn-ac corn-jc corn-btn-max corn-btn-primary"
      type="button"
    >
      @CornUI
    </button>
    <button
      class="corn-btn corn-if corn-ac corn-jc corn-btn-big corn-btn-primary"
      type="button"
    >
      @CornUI
    </button>
    <button
      class="corn-btn corn-if corn-ac corn-jc corn-btn-large corn-btn-primary"
      type="button"
    >
      @CornUI
    </button>
    <button
      class="corn-btn corn-if corn-ac corn-jc corn-btn-middle corn-btn-primary"
      type="button"
    >
      @CornUI
    </button>
    <button
      class="corn-btn corn-if corn-ac corn-jc corn-btn-small corn-btn-primary"
      type="button"
    >
      @CornUI
    </button>
    <button
      class="corn-btn corn-if corn-ac corn-jc corn-btn-mini corn-btn-primary"
      type="button"
    >
      @CornUI
    </button>
  </div>
</div>
`;

exports[`<Button /> type 1`] = `
<div>
  <div
    class=""
  >
    <button
      class="corn-btn corn-if corn-ac corn-jc corn-btn-middle corn-btn-error"
      type="button"
    >
      @CornUI
    </button>
    <button
      class="corn-btn corn-if corn-ac corn-jc corn-btn-middle corn-btn-link"
      type="button"
    >
      @CornUI
    </button>
    <button
      class="corn-btn corn-if corn-ac corn-jc corn-btn-middle corn-btn-pop"
      type="button"
    >
      @CornUI
    </button>
    <button
      class="corn-btn corn-if corn-ac corn-jc corn-btn-middle corn-btn-primary"
      type="button"
    >
      @CornUI
    </button>
    <button
      class="corn-btn corn-if corn-ac corn-jc corn-btn-middle corn-btn-warn"
      type="button"
    >
      @CornUI
    </button>
  </div>
</div>
`;

exports[`<Button /> 固定宽带 1`] = `
<div>
  <button
    class="corn-btn corn-if corn-ac corn-jc corn-btn-middle corn-btn-primary corn-btn-auto"
    type="button"
  />
</div>
`;

exports[`<Button /> 基本使用 1`] = `
<div>
  <button
    class="corn-btn corn-if corn-ac corn-jc corn-btn-middle corn-btn-primary"
    type="button"
  />
</div>
`;

exports[`<Button /> 复杂标题：节点+文字 1`] = `
<div>
  <button
    class="corn-btn corn-if corn-ac corn-jc corn-btn-middle corn-btn-primary"
    type="button"
  >
    <div
      class=""
    >
      <span
        class="taro-text"
      >
        我是图标
      </span>
    </div>
    <span
      class="taro-text"
    >
      @CornUI
    </span>
  </button>
</div>
`;

exports[`<Button /> 标题：@CornUI 1`] = `
<div>
  <button
    class="corn-btn corn-if corn-ac corn-jc corn-btn-middle corn-btn-primary"
    type="button"
  >
    @CornUI
  </button>
</div>
`;
